### DISCUSSION TITLE
let attributes in same scope be set both inside and outside profile context

### DISCUSSION THREAD

**Discussion by notestaff** [nextflow-io/nextflow#3490] [Q&A]
Current docs [warn](https://www.nextflow.io/docs/latest/config.html#config-profiles): "When using the profiles feature in your config file, do NOT set attributes in the same scope both inside and outside a profiles context."  Could this restriction be lifted?  E.g. in the process scope some directives that don't vary with profile are most naturally defined within the process definition in the .nf file (e.g. the conda env), while other directives are most naturally defined in profiles (e.g. clusterOptions for different clusters).   Also, it's odd that a setting of one config value in process scope can be overridden by specifying settings of _other_ config values in the same scope a profile context elsewhere.

Even more confusing, ignoring this restriction seems to work fine most of the time.   If it's an error to set attributes in the same scope both inside and outside a profiles context, Nextflow should error out, rather than silently doing something undefined.

**Comment by pditommaso** [comment]
Unfortunately, this is a known limitation. As you said it works most of the time but there are tricky use cases that are not supported, this is why it's advisable to not use this approach.

**Reply by notestaff** [reply]
Thanks for confirming.

If this limitation is hard to lift, I'd suggest making Nextflow throw an error when this constraint is violated, or at least print a strong warning.   It's easy to break this constraint accidentally, and the resulting problems often won't be obvious.

**Reply by abhi18av** [reply]
Good point - the discussion https://github.com/nextflow-io/nextflow/issues/2723 aims to address the concerns of over-riding config fields as well as inheritance of these values once the planned evolution of config has been implemented.
